### ISSUE TITLE
Errors related to Octave running make on MacOS

### ISSUE THREAD

**Issue by ankur-gupta** [stevengj/nlopt#88]
I get the following errors running `make` on MacOS Sierra, Octave 4.0.3. I have `mkoctfile` installed. 

```
Scanning dependencies of target nlopt_optimize
[ 97%] Building CXX object octave/CMakeFiles/nlopt_optimize.dir/nlopt_optimize-oct.cc.o
In file included from /Users/ankur/toolbox/github/nlopt/octave/nlopt_optimize-oct.cc:23:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/oct.h:33:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/Matrix.h:30:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/mx-base.h:28:
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/MatrixType.h:36:1: error: variable has incomplete type
      'class OCTAVE_API'
MatrixType
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/MatrixType.h:35:1: note: forward declaration of 'OCTAVE_API'
OCTAVE_API
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/MatrixType.h:36:11: error: expected ';' after top level
      declarator
MatrixType
          ^
          ;
In file included from /Users/ankur/toolbox/github/nlopt/octave/nlopt_optimize-oct.cc:23:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/oct.h:33:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/Matrix.h:30:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/mx-base.h:32:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/boolMatrix.h:27:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/Array.h:35:
In file included from /usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/dim-vector.h:33:
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:30:52: error: expected function body after
      function declarator
extern void liboctave_fatal (const char *fmt, ...) GCC_ATTR_NORETURN;
                                                   ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:32:76: error: expected function body after
      function declarator
extern void liboctave_fatal_with_id (const char *id, const char *fmt, ...) GCC_ATTR_NORETURN;
                                                                           ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:48:1: error: unknown type name 'CRUFT_API'
CRUFT_API extern liboctave_error_handler current_liboctave_error_handler;
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:48:11: error: expected unqualified-id
CRUFT_API extern liboctave_error_handler current_liboctave_error_handler;
          ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:50:1: error: unknown type name 'CRUFT_API'
CRUFT_API extern liboctave_error_with_id_handler current_liboctave_error_with_id_handler;
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:50:11: error: expected unqualified-id
CRUFT_API extern liboctave_error_with_id_handler current_liboctave_error_with_id_handler;
          ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:52:1: error: unknown type name 'CRUFT_API'
CRUFT_API extern liboctave_warning_handler current_liboctave_warning_handler;
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:52:11: error: expected unqualified-id
CRUFT_API extern liboctave_warning_handler current_liboctave_warning_handler;
          ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:54:1: error: unknown type name 'CRUFT_API'
CRUFT_API extern liboctave_warning_with_id_handler current_liboctave_warning_with_id_handler;
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:54:11: error: expected unqualified-id
CRUFT_API extern liboctave_warning_with_id_handler current_liboctave_warning_with_id_handler;
          ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:56:1: error: unknown type name 'CRUFT_API'
CRUFT_API extern void set_liboctave_error_handler (liboctave_error_handler f);
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:56:11: error: expected unqualified-id
CRUFT_API extern void set_liboctave_error_handler (liboctave_error_handler f);
          ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:58:1: error: unknown type name 'CRUFT_API'
CRUFT_API extern void set_liboctave_error_with_id_handler (liboctave_error_with_id_handler f);
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:58:11: error: expected unqualified-id
CRUFT_API extern void set_liboctave_error_with_id_handler (liboctave_error_with_id_handler f);
          ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:60:1: error: unknown type name 'CRUFT_API'
CRUFT_API extern void set_liboctave_warning_handler (liboctave_warning_handler f);
^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:60:11: error: expected unqualified-id
CRUFT_API extern void set_liboctave_warning_handler (liboctave_warning_handler f);
          ^
/usr/local/Cellar/octave/4.0.3_4/include/octave-4.0.3/octave/lo-error.h:62:1: error: unknown type name 'CRUFT_API'
CRUFT_API extern void set_liboctave_warning_with_id_handler (liboctave_warning_with_id_handler f);
^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [octave/CMakeFiles/nlopt_optimize.dir/nlopt_optimize-oct.cc.o] Error 1
make[1]: *** [octave/CMakeFiles/nlopt_optimize.dir/all] Error 2
make: *** [all] Error 2
```

I am interested in the Octave interface to NLopt.

**Comment by jslee02** [comment]
My two cents: I think this is because `oct.h` includes octave/config.h using angle brackets (see [line 31](http://octave.org/doxygen/4.0/d2/d3c/oct_8h_source.html)) so that it includes `nlopt/config.h` instead of `octave/config.h`. The ideal solution would be fixing it in the upstream by replacing the angle brackets with quotes. In the meantime, we could explicitly include `octave/config.h` at [here](https://github.com/stevengj/nlopt/blob/master/octave/nlopt_optimize-oct.cc#L22) as an workaround as:
```cpp
#include <octave/config.h>
#include <octave/oct.h>
#include <octave/oct-map.h>
#include <octave/ov.h>
#include <math.h>
#include <stdio.h>
```
One caveat is that this workaround causes warnings by including `octave/config.h` twice (sadly `octave/config.h` doesn't have header guard).

Maybe @stevengj has a better solution.

**Comment by ankur-gupta** [comment]
Thank you for the quick reply. I can try this out and see if it works.

**Comment by ankur-gupta** [comment]
Thanks @jslee02. That fixed the error for me for Octave 4.0.0.

**Comment by stevengj** [comment]
Can someone file an issue with Octave?

**Comment by stevengj** [comment]
Closed by #94.
